File: 2196-create-binary-tree-from-descriptions/2196-create-binary-tree-from-descriptions.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def createBinaryTree(self, descriptions: List[List[int]]) -> Optional[TreeNode]:
        # Step 1: Organize data
        all_nodes, children = set(), set()
        parents_to_children = defaultdict(list)
        
        for parent, child, is_left in descriptions:
            # Store child information under parent node
            parents_to_children[parent].append((child, is_left))
            
            all_nodes.add(parent)
            all_nodes.add(child)
            children.add(child)
            
        # Step 2: Find the root
        root_val = (all_nodes - children).pop()
        
        # Step 3 & 4: Build the tree using DFS
        def dfs(val: int) -> TreeNode:
            # Create new TreeNode for current value
            node = TreeNode(val)
            
            # If current node has children, recursively build them
            if val in parents_to_children:
                for child, is_left in parents_to_children[val]:
                    # Attach child node based on is_left flag
                    if is_left:
                        node.left = dfs(child)
                    else:
                        node.right = dfs(child)
                        
            return node
        
        return dfs(root_val)

File: 2196-create-binary-tree-from-descriptions/2196-create-binary-tree-from-descriptions.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* createBinaryTree(vector<vector<int>>& descriptions) {
        // Maps values to TreeNode pointers
        unordered_map<int, TreeNode*> nodeToMap;
        
        // Store values which are children in the descriptions
        unordered_set<int> children;
        
        // Iterate through descriptions to create nodes and set up tree structure
        for (vector<int>& d : descriptions){
            // Extract parent value, child value, and whether it's a 
            // left child (1) or right child (0)
            int parent = d[0], child = d[1], isLeft = d[2];
            
            // Create parent and child nodes if not already created
            if (!nodeToMap.contains(parent))
                nodeToMap[parent] = new TreeNode(parent);
            if (!nodeToMap.contains(child))
                nodeToMap[child] = new TreeNode(child);
            
            // Attach child node to parent's left or right branch
            if (isLeft)
                nodeToMap[parent] -> left = nodeToMap[child];
            else
                nodeToMap[parent] -> right = nodeToMap[child];
            
            // Mark child as a child in the set
            children.insert(child);
        }
        
        // Find and return the root node
        for (auto& [_, node] : nodeToMap){
            // Root node found
            if (!children.contains(node -> val))
                return node;
        }
        
        // Should not occur according to problem statement
        return nullptr;
    }
};